### PR TITLE
1406 change the signature of the function resolveVariables and fill t…

### LINKED ISSUE #1406: Generate with CRUD devon4j Server CRUD REST services fails with any input
### Expected behavior

As a CobiGen Eclipse user, I want to generate a CRUD REST service artifact.

### Actual behavior
When generating from an input entity or openAPI file, following merge error occurs:

`Unable to merge the patch generated by template D:\devon_ide\workspaces\junit-workspace\CobiGen_Templates\src\main\templates\crud_openapi_java_server_app\templates\java\${variables.rootPackage}\${variables.component#lower_case}\service\api\rest\${variables.component#cap_first}RestService.java.ftl into file C:\Users\username\AppData\Local\Temp\cobigen-12921793936326071041\src\main\java\sampledatamanagement\service\api\rest\SampledatamanagementRestService.java: The syntax of the base file is invalid. Error in line: 1 / column: 9: syntax error @[1,9]
Unable to merge the patch generated by template D:\devon_ide\workspaces\junit-workspace\CobiGen_Templates\src\main\templates\crud_openapi_java_server_app\templates\java\${variables.rootPackage}\${variables.component#lower_case}\service\api\rest\${variables.component#cap_first}RestService.java.ftl into file C:\Users\username\AppData\Local\Temp\cobigen-12921793936326071041\src\main\java\sampledatamanagement\service\api\rest\SampledatamanagementRestService.java: The syntax of the base file is invalid. Error in line: 1 / column: 9: syntax error @[1,9]
  com.devonfw.cobigen.api.exception.MergeException: Unable to merge the patch generated by template D:\devon_ide\workspaces\junit-workspace\CobiGen_Templates\src\main\templates\crud_openapi_java_server_app\templates\java\${variables.rootPackage}\${variables.component#lower_case}\service\api\rest\${variables.component#cap_first}RestService.java.ftl into file C:\Users\username\AppData\Local\Temp\cobigen-12921793936326071041\src\main\java\sampledatamanagement\service\api\rest\SampledatamanagementRestService.java: The syntax of the base file is invalid. Error in line: 1 / column: 9: syntax error @[1,9]
  	at com.devonfw.cobigen.impl.generator.GenerationProcessorImpl.generate(GenerationProcessorImpl.java:525)
  	at com.devonfw.cobigen.impl.generator.GenerationProcessorImpl.generate(GenerationProcessorImpl.java:212)
  	at com.devonfw.cobigen.impl.generator.CobiGenImpl.generate(CobiGenImpl.java:99)
  	at com.devonfw.cobigen.impl.generator.CobiGenImpl.generate(CobiGenImpl.java:82)
  	at com.devonfw.cobigen.eclipse.generator.CobiGenWrapper.generate(CobiGenWrapper.java:215)
  	at com.devonfw.cobigen.eclipse.wizard.generate.control.GenerateSelectionJob.performGeneration(GenerateSelectionJob.java:179)
  	at com.devonfw.cobigen.eclipse.wizard.generate.control.GenerateSelectionJob.run(GenerateSelectionJob.java:78)
  	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:122)
  Caused by: com.thoughtworks.qdox.parser.ParseException: syntax error @[1,9]
  	at com.thoughtworks.qdox.parser.impl.Parser.yyerror(Parser.java:1790)
  	at com.thoughtworks.qdox.parser.impl.Parser.yyparse(Parser.java:1912)
  	at com.thoughtworks.qdox.parser.impl.Parser.parse(Parser.java:1771)
  	at com.thoughtworks.qdox.library.SourceLibrary.parse(SourceLibrary.java:232)
  	at com.thoughtworks.qdox.library.SourceLibrary.parse(SourceLibrary.java:190)
  	at com.thoughtworks.qdox.library.SourceLibrary.addSource(SourceLibrary.java:94)
  	at com.thoughtworks.qdox.library.SourceLibrary.addSource(SourceLibrary.java:89)
  	at com.devonfw.cobigen.javaplugin.merger.libextension.ModifyableClassLibraryBuilder.addSource(ModifyableClassLibraryBuilder.java:42)
  	at com.devonfw.cobigen.javaplugin.inputreader.JavaParserUtil.getFirstJavaClass(JavaParserUtil.java:76)
  	at com.devonfw.cobigen.javaplugin.inputreader.JavaParserUtil.getFirstJavaClass(JavaParserUtil.java:44)
  	at com.devonfw.cobigen.javaplugin.merger.JavaMerger.merge(JavaMerger.java:76)
  	at com.devonfw.cobigen.impl.generator.GenerationProcessorImpl.generate(GenerationProcessorImpl.java:510)
  	... 7 more
`

### Steps to reproduce (bug) / Use Case of feature request (enhancement)

1. Create a simple project with an example OpenAPI file
2. Run CobiGen Generate
3. Choose f.e. CRUD REST services

### Affected version:

*   CobiGen Eclipse Plug-in	2021.8.3
*   CobiGen - OpenAPI Plug-in	2021.8.2
*   CobiGen - Java Plug-in	2021.8.2

### Acceptance Criteria

Sprint2:

* [ ] need for configuration adaption to make it possible to declare variableAssignments to be mandatory causing an intermediate error on missing variable. 
* [ ] please add website link into error message in case of openapi missing variables to help the user to help himself
* [ ] documentation update

### PR DESCRIPTION
**Pull request by LarsReinken**
…he report in OpenAPIMatcher

Adresses/Fixes #1406 .

Implements

* add a warning to the generationReport, if a required templateVariable isn't set

### CODE REVIEW

**Review by maybeec — APPROVED** [review]

**Review by maybeec — CHANGES_REQUESTED** [review]
You properly changed the signature, but I don't think this will fix the issue as the user will not take any notice from your changes.

File: cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/main/java/com/devonfw/cobigen/openapiplugin/matcher/OpenAPIMatcher.java [hunk 1]
```diff
@@ -141,12 +143,15 @@ public boolean matches(MatcherTo matcher) {
...
-  private String getExtendedProperty(Map<String, Object> extensionProperties, String key) {
+  private String getExtendedProperty(Map<String, Object> extensionProperties, String key, GenerationReportTo report) {
 
     Map<String, Object> properties = extensionProperties;
     if (properties.containsKey(key)) {
       return properties.get(key).toString();
     } else {
+      String warningMessage = "The property " + key
+          + " was requested in a variable assignment although the input does not provide this property. Setting it to empty";
+      report.addWarning(warningMessage);
```
Comment (maybeec) [inline cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/main/java/com/devonfw/cobigen/openapiplugin/matcher/OpenAPIMatcher.java:154]: Logging the warning is good, but not sure, how this will be shown to the user. Cannot see any code.
Also, it's not handled that this leads to a hard stop on the generation. Possibly you need to mark the variableAssignment to be mandatory or something like that and if so report an error instead of an warning and stop generation reporting the actual failure.

**Review by maybeec — CHANGES_REQUESTED** [review]

File: cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/test/java/com/devonfw/cobigen/openapiplugin/integrationtest/InputReaderMatcherTest.java [hunk 2]
```diff
@@ -253,17 +254,9 @@ public void testVariableAssignment_noAttributeFound() throws Exception {
...
-        .hasContent("testingAttributeTable");
-
-    template = findTemplate(cobigen, inputObjects.get(1), templateName);
-    targetFolder = this.tmpFolder.newFolder();
-    report = cobigen.generate(inputObjects.get(1), template, targetFolder.toPath());
-    assertThat(report).isSuccessful();
+    assertThat(report).containsException(CobiGenRuntimeException.class);
 
-    assertThat(targetFolder.toPath().resolve("testVariableAssignment_attribute.txt").toFile()).exists().hasContent("");
+    assertThat(targetFolder.toPath().resolve("testVariableAssignment_attribute.txt").toFile()).doesNotExist();
```
Comment (maybeec) [inline cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/test/java/com/devonfw/cobigen/openapiplugin/integrationtest/InputReaderMatcherTest.java:259]: Why is the test result changed? It's an existing good case, which should not be a failure after this change, shouldnt it?

**Review by maybeec — CHANGES_REQUESTED** [review]

File: cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/test/java/com/devonfw/cobigen/openapiplugin/unittest/matcher/OpenAPIMatcherTest.java [hunk 3]
```diff
@@ -73,4 +78,54 @@ public void testInvalidComponentDefMatching() {
...
+
+    matcher.resolveVariables(new MatcherTo("element", "ComponentDef", componentDef), va, report);
+    assertThat(report.getWarnings().size()).isEqualTo(0);
+  }
+
+  /**
+   * Test variable missing
+   */
+  @Test
+  public void testMissingXRootPackageVariable() {
```
Comment (jan-vcapgemini) [inline cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/test/java/com/devonfw/cobigen/openapiplugin/unittest/matcher/OpenAPIMatcherTest.java:105]: Needs a more detailed description (rootpackage).

File: cobigen/cobigen-core-api/src/main/java/com/devonfw/cobigen/api/to/VariableAssignmentTo.java [hunk 4]
```diff
@@ -21,7 +21,28 @@
...
+   */
+  protected boolean mandatory;
+
+  /**
+   * Constructor to create a {@link VariableAssignment} for a concrete string value. Requiredness as boolean.
+   *
+   * @param type Type of the variable assignment, interpreted by the plug-ins
+   * @param varName variable name
+   * @param value concrete string value
+   * @author mbrunnli (15.04.2013)
```
Comment (jan-vcapgemini) [inline cobigen/cobigen-core-api/src/main/java/com/devonfw/cobigen/api/to/VariableAssignmentTo.java:34]: change author and date?

File: cobigen/cobigen-core-api/src/main/java/com/devonfw/cobigen/api/to/VariableAssignmentTo.java [hunk 5]
```diff
@@ -21,7 +21,28 @@
   protected String value;
 
   /**
-   * Constructor to create a {@link VariableAssignmentTo} for a concrete string value
+   * True if the value is required, false if not
+   */
+  protected boolean mandatory;
+
+  /**
+   * Constructor to create a {@link VariableAssignment} for a concrete string value. Requiredness as boolean.
```
Comment (jan-vcapgemini) [inline cobigen/cobigen-core-api/src/main/java/com/devonfw/cobigen/api/to/VariableAssignmentTo.java:29]: replace requiredness with mandatory

File: cobigen/cobigen-core-api/src/test/java/com/devonfw/cobigen/api/matchers/VariableAssignmentToMatcher.java [hunk 6]
```diff
@@ -29,27 +29,34 @@
    */
   private Matcher<String> value;
 
+  /**
+   * Matcher for the requiredness of the variable
```
Comment (jan-vcapgemini) [inline cobigen/cobigen-core-api/src/test/java/com/devonfw/cobigen/api/matchers/VariableAssignmentToMatcher.java:33]: replace requiredness with mandatory

File: cobigen/cobigen-core-api/src/test/java/com/devonfw/cobigen/api/matchers/VariableAssignmentToMatcher.java [hunk 7]
```diff
@@ -29,27 +29,34 @@
...
+   */
+  private Matcher<Boolean> mandatory;
+
   /**
    * Creates a new {@link VariableAssignmentToMatcher} with the given attribute sub matchers
    *
    * @param typeMatcher matcher for the type of the variable assignment, interpreted by the plug-ins
    * @param varNameMatcher matcher for the variable's name
    * @param valueMatcher matcher for the value
+   * @param mandatoryMatcher matcher for the requiredness
```
Comment (jan-vcapgemini) [inline cobigen/cobigen-core-api/src/test/java/com/devonfw/cobigen/api/matchers/VariableAssignmentToMatcher.java:43]: replace requiredness with mandatory

**Review by jan-vcapgemini — CHANGES_REQUESTED** [review]

File: cobigen/cobigen-core-api/src/main/java/com/devonfw/cobigen/api/to/VariableAssignmentTo.java (hunk 4 above)
Comment (maybeec) [reply cobigen/cobigen-core-api/src/main/java/com/devonfw/cobigen/api/to/VariableAssignmentTo.java:34]: We can even remove it. It's legacy from old code formatting

File: cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/test/java/com/devonfw/cobigen/openapiplugin/integrationtest/InputReaderMatcherTest.java (hunk 2 above)
Comment (LarsReinken) [reply cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/test/java/com/devonfw/cobigen/openapiplugin/integrationtest/InputReaderMatcherTest.java]: you are right, the report should just contain warnings.

File: cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/test/java/com/devonfw/cobigen/openapiplugin/integrationtest/InputReaderMatcherTest.java [hunk 8]
```diff
@@ -261,9 +262,8 @@ public void testVariableAssignment_noAttributeFound() throws Exception {
     template = findTemplate(cobigen, inputObjects.get(1), templateName);
     targetFolder = this.tmpFolder.newFolder();
     report = cobigen.generate(inputObjects.get(1), template, targetFolder.toPath());
+    assertThat(report.hasWarnings());
     assertThat(report).isSuccessful();
-
-    assertThat(targetFolder.toPath().resolve("testVariableAssignment_attribute.txt").toFile()).exists().hasContent("");
```
Comment (jan-vcapgemini) [inline cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/test/java/com/devonfw/cobigen/openapiplugin/integrationtest/InputReaderMatcherTest.java:266]: Why was this removed?

**Review by jan-vcapgemini — CHANGES_REQUESTED** [review]

**Review by jan-vcapgemini — DISMISSED** [review]

**Review by MikeSchumacherCapgemini — DISMISSED** [review]

File: cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/test/java/com/devonfw/cobigen/openapiplugin/unittest/matcher/OpenAPIMatcherTest.java [hunk 9]
```diff
@@ -73,4 +78,74 @@ public void testInvalidComponentDefMatching() {
...
+            + "although the input does not provide this property. Setting it to empty");
+
+    List<VariableAssignmentTo> vaMandatoryXRootPackage = new ArrayList<>();
+    vaMandatoryXRootPackage.add(new VariableAssignmentTo("extension", "rootPackage", "x-rootpackage", true));
+
+    matcher.resolveVariables(new MatcherTo("element", "ComponentDef", componentDef), vaMandatoryXRootPackage, report);
+    assertThat(report.getErrors().get(0).getMessage())
+        .containsSequence("The property x-rootpackage was required in a variable assignment "
+            + "although the input does not provide this property. "
+            + "Please add the required attribute in your input file or set the \"mandatory\" attribute to \"false\". ");
```
Comment (maybeec) [inline cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/test/java/com/devonfw/cobigen/openapiplugin/unittest/matcher/OpenAPIMatcherTest.java:128]: I feel we should externalize this string and reuse it By making use of string formatter. But the string already occurred quite some time and would make small changes nasty to be done

**Review by maybeec — CHANGES_REQUESTED** [review]

File: cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/test/java/com/devonfw/cobigen/openapiplugin/unittest/matcher/OpenAPIMatcherTest.java (hunk 9 above)
Comment (MikeSchumacherCapgemini) [reply cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/test/java/com/devonfw/cobigen/openapiplugin/unittest/matcher/OpenAPIMatcherTest.java]: Pulled the error and warning message into a method in the Constants class, passing the value that was/wan't set to mandatory as a parameter

**Review by maybeec — APPROVED** [review]

File: cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/main/java/com/devonfw/cobigen/openapiplugin/util/constants/Constants.java [hunk 10]
```diff
@@ -18,4 +18,16 @@
 
   public static final String SCHEMA = "schema";
 
+  public static String getMandatoryErrorMessage(String property) {
```
Comment (jan-vcapgemini) [inline cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/main/java/com/devonfw/cobigen/openapiplugin/util/constants/Constants.java:21]: Javadoc missing.

File: cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/main/java/com/devonfw/cobigen/openapiplugin/util/constants/Constants.java [hunk 11]
```diff
@@ -18,4 +18,16 @@
...
+
+    return "The property " + property.toLowerCase()
+        + " was required in a variable assignment although the input does not provide this property. "
+        + "Please add the required attribute in your input file or set the \"mandatory\" attribute to \"false\". ";
+  }
+
+  public static String getMandatoryWarning(String property) {
```
Comment (jan-vcapgemini) [inline cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/main/java/com/devonfw/cobigen/openapiplugin/util/constants/Constants.java:28]: Javadoc missing.

File: cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/main/java/com/devonfw/cobigen/openapiplugin/matcher/OpenAPIMatcher.java [hunk 12]
```diff
@@ -140,17 +150,15 @@ public boolean matches(MatcherTo matcher) {
...
 
     Map<String, Object> properties = extensionProperties;
     if (properties.containsKey(key)) {
       return properties.get(key).toString();
     } else {
-      LOG.warn(
-          "The property {} was requested in a variable assignment although the input does not provide this property. Setting it to empty",
-          key);
-      return "";
+      throw new NoSuchFieldException("couldn't find property: " + key);
```
Comment (jan-vcapgemini) [inline cobigen-plugins/cobigen-openapiplugin-parent/cobigen-openapiplugin/src/main/java/com/devonfw/cobigen/openapiplugin/matcher/OpenAPIMatcher.java:161]: Shouldn't this message be a bit more detailed?
